release/models/system/openconfig-system-grpc-tunnel.yang

```diff
@@ -0,0 +1,348 @@
+      enum OPENFLOW {
+        value 6653;
+        description
+          "OpenFlow protocol target type";
+      }
+      enum GNMI_GNOI {
+        value 9339;
+        description
+          "gNMI/gNOI target type";
+      }
+      enum GRIBI {
+        value 9340;
+        description
+          "gRIBI target type";
+      }
+      enum P4RT {
+        value 9559;
+        description
+          "P4 Runtime target type";
+      }
+    }
+    description
+      "Sets the TCP based protocols to be tunneled over the gRPC tunnel";
+    reference
+      "OpenConfig gRPC Tunnel:
+      https://github.com/openconfig/grpctunnel/blob/master/proto/tunnel/tunnel.proto";
+  }
+
+  grouping tunnel-server-config {
+    description
+      "Configuration and state parameters corresponding to an individual
+      gRPC tunnel server";
+
+    leaf name {
+      type string;
+      description
+        "The name of the gRPC tunnel server instance";
+    }
+
+    leaf enable {
+      type boolean;
+      description
+        "When set to true, the tunnel server endpoint is enabled and an
+        outbound connection is initiated";
+    }
+
+    leaf host {
+      type oc-inet:host;
+      description
+        "The IPv4/IPv6 address or hostname of the tunnel server";
+    }
+
+    leaf port {
+      type oc-inet:port-number;
+      description
+        "The port number of the tunnel server";
+    }
+
+    leaf transport-security {
+      type boolean;
+      default true;
+      description
+        "Use gRPC transport security (TLS).  Enabled by default.  This
+        leaf allows transport security to be disabled for use cases that
+        are not supported, such as lab testing.";
+    }
+
+    leaf certificate-id {
+      type string;
+      description
+        "Name of the certificate that is associated with client (e.g.
+        mutual authentication).  The certificate ID is provisioned
+        through other interfaces to the device, such as the gNOI
+        certificate management service.  If this leaf is provisioned and
+        transport-security is enabled, mutual authentication is implied.
+        If this leaf is not provisioned and transport-security is
+        enabled, then the client will only validate the server
+        certificate (one-way authentication).";
+    }
+
+    leaf source-address {
+      type oc-inet:ip-address;
+      description
+        "Source IPv4/IPv6 address to use in communications to the tunnel
+        server";
+    }
+
+    leaf network-instance {
+      type oc-netinst:network-instance-ref;
+      description
+        "The network instance to initiate outbound communications to the
+        tunnel server.  When unspecified, the DEFAULT_INSTANCE should be
+        used.";
+    }
+
+    leaf dscp {
+      type oc-inet:dscp;
+      description
+        "DSCP marking of packets generated by the gRPC tunnel client";
+    }
+
+    leaf-list target {
+      type union {
+        type target-type;
+        type string;
```

Comment:
If there is a string here, what does it mean? In the cases above it seems to imply the TCP ports that will be plumbed through -- in the case where there is just a string wouldn't it be better to have `target-type` rather be an identity and then expand on this with the protocols that are not defined in the base model?

Reply:
This actually accommodates @mirovr's query which is valid in that grpc-tunnel is meant to tunnel TCP applications however per the grpc-tunnel specification there is only a small number of TCP applications defined (what is brought over to here into the typedef `target-type`) - in theory you could target any TCP application that may or may not be worthy of being defined in OpenConfig modules as enums or identities.

We could move this to an identityref and add identities (instead of enum variants) each time but this is churn that may not scale and would not accomodate any non-vendor neutral TCP applications.

Open to suggestions - this gives a looser option for users where the strict enum/identity method is precise and limiting